packages/cli/cli/src/cli-context/upgrade-utils/getFernUpgradeMessage.ts

```diff
@@ -0,0 +1,162 @@
+import boxen from "boxen";
+import chalk from "chalk";
+import { FernUpgradeInfo } from "../CliContext";
+import { CliEnvironment } from "../CliEnvironment";
+import { FernGeneratorUpgradeInfo } from "./getGeneratorVersions";
+
+export function getFernUpgradeMessage({
```

Comment:
nit: Not functionally any different, but more of a code style suggestion -

It looks like we do a fair amount of work and compose a `message` that isn't always used (i.e. we return `undefined` at the end and the `message` is dropped altogether).

We could rearrange this a bit so that we first detect whether or not we need to do anything, and return early if so. That way we can drop the `upgradeAvailable` variable entirely.

```typescript
if (!hasUpgrade(upgradeInfo)) {
  return undefined;
}
...

function hasUpgrade(upgradeInfo: FernUpgradeInfo): boolean {
  return upgradeInfo.cliUpgradeInfo?.isUpgradeAvailable || (upgradeInfo.generatorUpgradeInfo.filter((gui) => gui.isUpgradeAvailable)).length > 0
}
```

---

Alternatively, if we find ourselves repeating the same functions to determine whether or not there is an upgrade for CLI vs. generators, we could make `FernUpgradeInfo` a class and give it a method like:

```typescript
if (!upgradeInfo.hasUpgrade()) {
  ...
}
if (upgradeInfo.hasGeneratorUpgrade()) {
  ...
}
...
```